File: admin_frontend/src/lib/api/chat-channels.ts
import { base } from '$app/paths';
import { PREF_KEYS } from '$lib/preferences/keys';
import { apiRequest } from './client';

export type ChatChannelRow = {
  id: number;
  name: string;
  type: string;
  character_id: string;
  user_id: number;
  description?: string;
  created_at: string;
  last_message_at: string | null;
  last_deleted: number;
  character?: { id: string; name: string };
  capabilities?: unknown;
  is_lowest_id_channel?: boolean;
  photo_data_url?: string | null;
  thumbnail_mtime_ns?: number;
};

/** One item in ``messages.history`` / admin list (text, audio, …). */
export type ChatMessageContentItem = {
  content_type: string;
  body: string;
  metadata?: Record<string, unknown>;
};

export type AgentUsageTotals = {
  input_tokens?: number;
  output_tokens?: number;
  total_tokens?: number;
  cached_input_tokens?: number;
  reasoning_tokens?: number;
};

export type AgentToolCall = {
  id?: string;
  name?: string;
  status?: string;
  elapsed_ms?: number;
  /** Single-line JSON (or string) of invocation parameters. */
  args?: string;
  /** Tool return payload (truncated server-side when persisted). */
  result?: string;
  error?: string;
};

export type AgentCostSummary = {
  currency?: string;
  estimated_total?: number;
  pricing_available?: boolean;
  reason?: string;
};

export type AgentMessageMetadata = {
  status?: string;
  current_step?: string | null;
  last_event?: string;
  reply_id?: string;
  elapsed_ms?: number;
  error?: {
    message?: string;
    code?: string;
    node?: string;
  };
  usage_total?: AgentUsageTotals;
  cost?: AgentCostSummary;
  tools?: AgentToolCall[];
};

/** Normalized history row from ``_sync_history`` (admin chat Messages tab). */
export type ChatHistoryMessage = {
  /** Protocol / external message id (UUID hex, etc.). */
  id: string;
  message_pk: number;
  channel_id: number;
  sender_type: string;
  sender_id: string;
  created_at: string;
  content: ChatMessageContentItem[];
  metadata?: Record<string, unknown> & { agent?: AgentMessageMetadata };
};

export type ChatChannelPayload = {
  name: string;
  character_id: string;
  description: string;
};

const apiBase = `${base}/api`;

export function historyMessageText(m: ChatHistoryMessage): string {
  return m.content
    .filter((c) => c.content_type === 'text')
    .map((c) => c.body)
    .filter(Boolean)
    .join('\n')
    .trim();
}

export function historyMessageFirstAudio(m: ChatHistoryMessage): ChatMessageContentItem | null {
  return m.content.find((c) => c.content_type === 'audio') ?? null;
}

/** Parse slot index from ``message_attachment:<id>:<slot>`` body. */
export function parseMessageAttachmentSlot(body: string): number | null {
  if (!body.startsWith('message_attachment:')) return null;
  const rest = body.slice('message_attachment:'.length);
  const idx = rest.lastIndexOf(':');
  if (idx < 0) return null;
  const n = Number.parseInt(rest.slice(idx + 1), 10);
  return Number.isNaN(n) ? null : n;
}

export async function listChatChannels() {
  return apiRequest<ChatChannelRow[]>('/chat-channels');
}

export async function createChatChannel(payload: ChatChannelPayload) {
  return apiRequest<ChatChannelRow>('/chat-channels', { method: 'POST', body: payload });
}

export async function updateChatChannel(channelId: number, payload: ChatChannelPayload) {
  return apiRequest<ChatChannelRow>(`/chat-channels/${channelId}`, {
    method: 'PATCH',
    body: payload
  });
}

export async function deleteChatChannel(channelId: number) {
  return apiRequest<number>(`/chat-channels/${channelId}`, { method: 'DELETE' });
}

export async function uploadChatChannelPhoto(channelId: number, dataUrl: string) {
  return apiRequest<null>(`/chat-channels/${channelId}/photo`, {
    method: 'POST',
    body: { data_url: dataUrl }
  });
}

export type ListChatMessagesOptions = {
  after?: string;
  afterId?: string;
  limit?: number;
  messagePks?: number[];
};

export async function listChatMessages(
  channelId: number,
  options: ListChatMessagesOptions = {}
) {
  const params = new URLSearchParams();
  const hasAfter = options.after !== undefined;
  const hasAfterId = options.afterId !== undefined;
  const hasMessagePks = (options.messagePks?.length ?? 0) > 0;
  if (hasMessagePks && (hasAfter || hasAfterId || options.limit !== undefined)) {
    throw new Error('messagePks cannot be combined with cursor options.');
  }
  if (hasAfter !== hasAfterId) {
    throw new Error('after and afterId must be provided together.');
  }
  if (hasAfter && options.limit === undefined) {
    throw new Error('limit is required with after and afterId.');
  }
  if (!hasAfter && !hasMessagePks && options.limit !== undefined) {
    throw new Error('limit is only supported with after and afterId.');
  }

  if (hasAfter && hasAfterId) {
    params.set('after', options.after!);
    params.set('after_id', options.afterId!);
    params.set('limit', String(options.limit));
  }
  if (hasMessagePks) {
    for (const pk of options.messagePks ?? []) {
      params.append('message_pk', String(pk));
    }
  }
  const query = params.toString();
  return apiRequest<ChatHistoryMessage[]>(
    `/chat-channels/${channelId}/messages${query ? `?${query}` : ''}`
  );
}

/** Proxy to workspace Hiro POST /invoke message_send — server must be running. */
export async function sendChatMessage(
  channelId: number,
  body: {
    text?: string;
    audio_base64?: string;
    audio_mime_type?: string;
    audio_duration_ms?: number;
    request_voice_reply?: boolean;
    use_knowledge?: boolean;
  }
) {
  return apiRequest<{ message_id: string; channel_id: number }>(
    `/chat-channels/${channelId}/messages/send`,
    { method: 'POST', body }
  );
}

/** Bulk-delete conversation messages — channel stays; bumps ``last_deleted`` on devices. */
export async function clearChatMessages(channelId: number) {
  return apiRequest<{ channel_id: number; last_deleted: number }>(
    `/chat-channels/${channelId}/messages/clear`,
    { method: 'POST' }
  );
}

/**
 * Load attachment bytes for `<audio>` — uses ``X-Hiro-Workspace`` (plain media
 * ``<audio src>`` cannot).
 */
export async function fetchChatMessageAttachmentBlob(
  channelId: number,
  externalMessageId: string,
  slot: number
): Promise<Blob> {
  const selectedWorkspace =
    typeof localStorage === 'undefined' ? null : localStorage.getItem(PREF_KEYS.selectedWorkspace);
  const headers = new Headers();
  if (selectedWorkspace) {
    headers.set('x-hiro-workspace', selectedWorkspace);
  }
  const enc = encodeURIComponent(externalMessageId);
  const response = await fetch(
    `${apiBase}/chat-channels/${channelId}/messages/by-external/${enc}/attachments/${slot}/media`,
    { headers }
  );
  if (!response.ok) {
    const err = await response.text().catch(() => '');
    throw new Error(err || `HTTP ${response.status}`);
  }
  return response.blob();
}

File: hiroserver/hirocli/src/hirocli/admin_svelte/routes/chat_channels.py
"""Chat channel admin routes."""

from __future__ import annotations

from functools import partial
from pathlib import Path
from tempfile import NamedTemporaryFile
from typing import Any

from fastapi import APIRouter, HTTPException, Query
from fastapi.responses import FileResponse
from starlette.concurrency import run_in_threadpool

from hirocli.admin.features.chat_channels.service import ChatChannelsService
from hirocli.admin_svelte.deps import SelectedWorkspaceIdDep
from hirocli.admin_svelte.photo_decode import _decode_photo_data_url
from hirocli.admin_svelte.result_payload import _api_from_result, envelope_failure
from hirocli.admin_svelte.schemas import ChatChannelMessageSendRequest, ChatChannelPhotoUploadRequest, ChatChannelSaveRequest

chat_channels_router = APIRouter()

_MAX_MESSAGE_PK_RESYNC = 16


@chat_channels_router.get("/chat-channels")
async def list_chat_channels(workspace_id: SelectedWorkspaceIdDep) -> dict[str, Any]:
    result = await run_in_threadpool(
        ChatChannelsService().list_channels,
        workspace_id,
    )
    payload = _api_from_result(result)
    payload["data"] = payload["data"] or []
    return payload


@chat_channels_router.post("/chat-channels")
async def create_chat_channel(
    body: ChatChannelSaveRequest,
    workspace_id: SelectedWorkspaceIdDep,
) -> dict[str, Any]:
    result = await run_in_threadpool(
        partial(
            ChatChannelsService().create_channel,
            workspace_id,
            name=body.name,
            character_id=body.character_id,
            description=body.description,
        )
    )
    return _api_from_result(result)


@chat_channels_router.patch("/chat-channels/{channel_id}")
async def update_chat_channel(
    channel_id: int,
    body: ChatChannelSaveRequest,
    workspace_id: SelectedWorkspaceIdDep,
) -> dict[str, Any]:
    result = await run_in_threadpool(
        partial(
            ChatChannelsService().update_channel,
            workspace_id,
            channel_id,
            name=body.name,
            character_id=body.character_id,
            description=body.description,
        )
    )
    return _api_from_result(result)


@chat_channels_router.delete("/chat-channels/{channel_id}")
async def delete_chat_channel(
    channel_id: int,
    workspace_id: SelectedWorkspaceIdDep,
) -> dict[str, Any]:
    result = await run_in_threadpool(
        ChatChannelsService().delete_channel,
        workspace_id,
        channel_id,
    )
    return _api_from_result(result)


@chat_channels_router.post("/chat-channels/{channel_id}/messages/clear")
async def clear_chat_channel_messages(
    channel_id: int,
    workspace_id: SelectedWorkspaceIdDep,
) -> dict[str, Any]:
    result = await run_in_threadpool(
        ChatChannelsService().clear_messages,
        workspace_id,
        channel_id,
    )
    return _api_from_result(result)


@chat_channels_router.post("/chat-channels/{channel_id}/messages/send")
async def send_chat_channel_message(
    channel_id: int,
    body: ChatChannelMessageSendRequest,
    workspace_id: SelectedWorkspaceIdDep,
) -> dict[str, Any]:
    """Proxy to workspace Hiro ``POST /invoke`` → ``message_send`` (live server)."""
    result = await run_in_threadpool(
        partial(
            ChatChannelsService().send_chat_message,
            workspace_id,
            channel_id,
            text=body.text,
            audio_base64=body.audio_base64,
            audio_mime_type=body.audio_mime_type,
            audio_duration_ms=body.audio_duration_ms,
            request_voice_reply=body.request_voice_reply,
            use_knowledge=body.use_knowledge,
        )
    )
    return _api_from_result(result)


@chat_channels_router.get("/chat-channels/{channel_id}/messages")
async def list_chat_channel_messages(
    channel_id: int,
    workspace_id: SelectedWorkspaceIdDep,
    after: str | None = None,
    after_id: str | None = None,
    limit: int | None = None,
    message_pk: list[int] | None = Query(default=None),
) -> dict[str, Any]:
    has_cursor_param = after is not None or after_id is not None
    has_message_pk = message_pk is not None and len(message_pk) > 0
    if has_message_pk and (has_cursor_param or limit is not None):
        raise HTTPException(
            status_code=400,
            detail="message_pk cannot be combined with cursor or limit parameters.",
        )
    if has_message_pk and len(message_pk) > _MAX_MESSAGE_PK_RESYNC:
        raise HTTPException(
            status_code=400,
            detail=f"message_pk is limited to {_MAX_MESSAGE_PK_RESYNC} values.",
        )
    if (after is None) != (after_id is None):
        raise HTTPException(
            status_code=400,
            detail="after and after_id must be provided together.",
        )
    if has_cursor_param and limit is None:
        raise HTTPException(
            status_code=400,
            detail="limit is required with after and after_id.",
        )
    if limit is not None and not has_cursor_param:
        raise HTTPException(
            status_code=400,
            detail="limit is only supported with after and after_id.",
        )
    if limit is not None and limit < 1:
        raise HTTPException(status_code=400, detail="limit must be positive.")

    result = await run_in_threadpool(
        partial(
            ChatChannelsService().list_messages_all,
            workspace_id,
            channel_id,
            after=after,
            after_id=after_id,
            limit=limit,
            message_pks=message_pk if has_message_pk else None,
        )
    )
    payload = _api_from_result(result)
    payload["data"] = payload["data"] or []
    return payload


@chat_channels_router.get(
    "/chat-channels/{channel_id}/messages/by-external/{external_message_id}/attachments/{slot}/media",
)
async def get_chat_message_attachment_media(
    channel_id: int,
    external_message_id: str,
    slot: int,
    workspace_id: SelectedWorkspaceIdDep,
) -> FileResponse:
    """Stream one message attachment for admin `<audio>` playback (workspace-scoped)."""
    if slot < 0:
        raise HTTPException(status_code=400, detail="Invalid attachment slot.")
    result = await run_in_threadpool(
        partial(
            ChatChannelsService().resolve_message_attachment_media,
            workspace_id,
            channel_id,
            external_message_id,
            slot,
        )
    )
    if not result.ok or result.data is None:
        raise HTTPException(
            status_code=404,
            detail=result.error or "Attachment not found.",
        )
    path_str, media_type = result.data
    return FileResponse(path_str, media_type=media_type)


@chat_channels_router.post("/chat-channels/{channel_id}/photo")
async def upload_chat_channel_photo(
    channel_id: int,
    body: ChatChannelPhotoUploadRequest,
    workspace_id: SelectedWorkspaceIdDep,
) -> dict[str, Any]:
    try:
        raw = _decode_photo_data_url(body.data_url)
    except ValueError as exc:
        return envelope_failure(str(exc))
    with NamedTemporaryFile(delete=False, suffix=".png") as tmp:
        tmp.write(raw)
        tmp_path = tmp.name
    try:
        result = await run_in_threadpool(
            ChatChannelsService().upload_channel_photo,
            workspace_id,
            channel_id,
            tmp_path,
        )
    finally:
        Path(tmp_path).unlink(missing_ok=True)
    return _api_from_result(result)
